### PR TITLE
adding proto files and adding component operate

### PR DESCRIPTION
**Pull request by yashjaind11**
# Odin version(x.y.z)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -3,7 +3,7 @@ install:
 	go mod download
 	go build .
 	sudo mv ./odin /usr/local/bin
-	odin version
+	odin --version
```
Comment (boppanasusanth) [inline Makefile:6]: why the change. `odin --version` doesn't work in the new cli

File: proto/gen/go/dream11/od/auth/v1/auth_grpc.pb.go [hunk 2]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
 // versions:
 // - protoc-gen-go-grpc v1.3.0
-// - protoc             v5.26.1
+// - protoc             v3.20.3
```
Comment (boppanasusanth) [inline proto/gen/go/dream11/od/auth/v1/auth_grpc.pb.go:4]: install the latest protoc version with `brew install protobuf` and generate the proto files again

File: proto/dream11/od/service/v1/service.proto [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+message ComponentProvisioningConfig {
+  string component_name = 1;
+  string deployment_type = 2;
+  google.protobuf.Struct params = 3;
+}
+
+message ProvisioningConfig {
+  repeated ComponentProvisioningConfig component_provisioning_config = 1;
+}
```
Comment (boppanasusanth) [inline proto/dream11/od/service/v1/service.proto:31]: add newline at the end of file

**Comment by boppanasusanth** [comment]
run `make lint` once. if you get an error saying `golangci-lint: command not found`, run `brew install golangci-lint`

File: cmd/operate/component.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+func init() {
+	operateComponentCmd.Flags().StringVar(&name, "name", "", "name of the component")
+	operateComponentCmd.Flags().StringVar(&serviceName, "service", "", " name of the service in which the component is deployed")
+	operateComponentCmd.Flags().StringVar(&env, "env", "", "name of the environment in which the service is deployed")
+	operateComponentCmd.Flags().StringVar(&operation, "operation", "", "name of the operation to performed on the component")
+	operateComponentCmd.Flags().StringVar(&options, "options", "{}", "options of the operation in JSON format")
+	operateComponentCmd.Flags().StringVar(&file, "file", "", "path of the file which contains the options for the operation in JSON format")
+	if err := operateComponentCmd.MarkFlagRequired("name"); err != nil {
+		log.Fatal("Error:  --env, --name, --operation, --service cannot be blank", err)
```
Comment (boppanasusanth) [inline cmd/operate/component.go:43]: Change the error here. The error is when marking a flag as required is failing. see create/environment.go for an example error message for the same

Comment (boppanasusanth) [inline cmd/operate/component.go:43]: change the error here. this error is not null when marking the flag as required is failing, not when user didn't set a value for the flag. check create environment for the error message to be used here

File: internal/service/component.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	if err != nil {
+		return err
+	}
+	client := operate.NewServiceServiceClient(conn)
+	stream, err := client.OperateService(*requestCtx, request)
+	if err != nil {
+		return err
+	}
+
+	log.Info("Operate Component in progress...")
```
Comment (boppanasusanth) [inline internal/service/component.go:28]: change the message here to `Operating component...` or `Starting component operation`. The message in the spinner may also be `Operate component in progress` and it doesn't look good for both to the same sentence

File: proto/dream11/od/service/v1/service.proto
Comment (boppanasusanth) [inline proto/dream11/od/service/v1/service.proto:1]: this proto file should be in od/dto

File: proto/dream11/od/operate/v1/operate.proto
Comment (boppanasusanth) [inline proto/dream11/od/operate/v1/operate.proto:1]: We make the proto files for the resource not the verb. So this should be named `component.proto` and should be in `od/component/v1`

File: proto/dream11/od/operate/v1/operate.proto [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  bool is_component_operation = 4;
+  string operation = 5;
+  google.protobuf.Struct config = 6;
+}
+
+message OperateServiceResponse {
+  ServiceStatus service_status = 1;
+  repeated ComponentStatus components_status = 2;
+  string message = 3;
+}
```
[the remark this answers is not in the document]
Comment (yashjaind11) [reply proto/dream11/od/operate/v1/operate.proto:65]: ok

File: proto/gen/go/dream11/od/auth/v1/auth_grpc.pb.go (hunk 2 above)
Comment (yashjaind11) [reply proto/gen/go/dream11/od/auth/v1/auth_grpc.pb.go:4]: done

File: pkg/file/file.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+// Write : write data to file
+func Write(filePath, data string, permission fs.FileMode) error {
+	byteData := []byte(data)
+	err := os.WriteFile(filePath, byteData, permission)
+
+	return err
+}
+
+func FindAndReadAllAllowedFormat(path string, allowedFormats []string) ([]byte, string, error) {
```
Comment (boppanasusanth) [inline pkg/file/file.go:26]: add comment above this exported function

File: pkg/util/util.go [hunk 8]
```diff
@@ -18,3 +19,15 @@ func IsIPAddress(address string) bool {
 	addr := net.ParseIP(address)
 	return addr != nil
 }
+
+// EmptyParameters get empty parameter list check if the map have any key with empty value if yes then return a comma seperated string of keys which have empty values
+func EmptyParameters(params map[string]string) string {
```
Comment (boppanasusanth) [inline pkg/util/util.go:24]: remove this function. this is not needed in the new cli

**Review by boppanasusanth — APPROVED** [review]

File: cmd/operate/env.go [hunk 9]
```diff
@@ -0,0 +1 @@
+package operate
```
Comment (dhavalmehta07) [inline cmd/operate/env.go:1]: why do we need this?

File: pkg/file/file.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
+package file
```
Comment (dhavalmehta07) [inline pkg/file/file.go:1]: should the package be `file` or `utils`?

File: pkg/file/file.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+	yamlPorvider "gopkg.in/yaml.v3"
+)
+
+// Read : read data from file
+func Read(filePath string) ([]byte, error) {
+	return os.ReadFile(filePath)
+}
+
+// Write : write data to file
+func Write(filePath, data string, permission fs.FileMode) error {
```
Comment (dhavalmehta07) [inline pkg/file/file.go:19]: is this being used?

File: cmd/operate/env.go (hunk 9 above)
Comment (yashjaind11) [reply cmd/operate/env.go:1]: this is for future development

File: pkg/file/file.go (hunk 10 above)
Comment (yashjaind11) [reply pkg/file/file.go:1]: util ...correcting this

File: pkg/file/file.go (hunk 11 above)
Comment (yashjaind11) [reply pkg/file/file.go:19]: nope ...just a util for future use

File: cmd/operate/env.go (hunk 9 above)
Comment (dhavalmehta07) [reply cmd/operate/env.go:1]: we dont have a env operation yet, even in the existing product. Lets remove it, we will add when we need it.

Comment (boppanasusanth) [reply cmd/operate/env.go:1]: @dhavalmehta07 we have env operations
